### PR TITLE
🔨 ⚡ 🦀 🔒 Make rust sandbox safe with crates

### PR DESCRIPTION
**Pull request by simonrainerson**
This adds a set of rust crates, under base.languages.rust.crates, and tooling to extend, override or replace that set. This set is built from fixed output derivations so rust will no longer access crates.io during build. It also simplifies the logic by treating external and internal crates the same. It also removes a limitation where users could not create a custom .cargo/config.toml file.

Removed the test for internal dependencies, because the use of
setupHooks the vendoring happens build time and the eval time test can't
verify anything.

The python script to generate nix expression is not perfect, but since it's a helper tool and not something that is actually needed it can be improved over time. Current problems:
 - It's slow: downloading the whole index even if you just need one crate. Attempted to use github's rest api instead but got rate limited when traversing to find deep and many dependencies.
 - It doesn't consider version requirements on dependencies.
 - It doesn't consider features, features requiring features etc.
Might be better to replace with something written in rust or a combination of `cargo search`, `cargo vendor` and `cargo tree`.

### CODE REVIEW

**Comment by abbec** [comment]
Would it be possible to make the dependencies normal dependencies that register a (singleton) setup hook that does the cargo setup? I would like it if the Rust crates could be normal dependencies (i.e buildInputs) just like in Python and c/c++.

Otherwise, this is awesome :)

**Comment by abbec** [comment]
> Would it be possible to make the dependencies normal dependencies that register a (singleton) setup hook that does the cargo setup? I would like it if the Rust crates could be normal dependencies (i.e buildInputs) just like in Python and c/c++.

Which I see now that it already does. But maybe then just move the logic to the setup hook in fetchCrate? In summary, do whatever Python derivations do w.r.t. PYTHONPATH 🥳

**Comment by simonrainerson** [comment]
> > Would it be possible to make the dependencies normal dependencies that register a (singleton) setup hook that does the cargo setup? I would like it if the Rust crates could be normal dependencies (i.e buildInputs) just like in Python and c/c++.
> 
> Which I see now that it already does. But maybe then just move the logic to the setup hook in fetchCrate? In summary, do whatever Python derivations do w.r.t. PYTHONPATH 🥳

Yes, just needs a little think of how to do it so it work for the shell in that case. When the setupHook executes NIX_BUILD_TOP and TMP and friends are for the derivation (I think, maybe has something to do with our use of mkShell) and when the shell runs those temp folders are yeeted. But it needed a bit of a rework anyway since we accommodate for producing multiple crate files from a mkLibrary.

File: rust/default.nix [hunk 1]
```diff
@@ -198,13 +189,16 @@ let
...
 
               for crate in target/package/*.crate; do
-                tar -xzf $crate -C $out/src/rust
+                tar -xzf $crate -C $out
+                echo "{\"files\":{},\"package\":\"$(sha256sum $crate | grep -E -o '^(\w*)')\"}" >$out/"$(basename "''${crate//.crate/}")"/.cargo-checksum.json
               done
               runHook postInstall
             '';
+
+            setupHook = ./crate-setup-hook.sh;
```
Comment (abbec) [inline rust/default.nix]: Should this be removed now when it is in `fetchCrate`?

Comment (abbec) [reply rust/default.nix]: The above comment is very very wrong

**Review by abbec — APPROVED** [review]

**Review by abbec — APPROVED** [review]

**Review by sakarias88 — APPROVED** [review]

**Pull request merged**